File: Questions/Easy/Parking Lot System/python/level.py
from typing import List

from models.vehicle import Vehicle
from parking_spot import ParkingSpot


class Level:
    def __init__(self, floor, capacity):
        self.floor = floor
        self.parking_spots: List[ParkingSpot] = [ParkingSpot(idx) for idx in range(capacity)]

    def park(self, vehicle: Vehicle) -> bool:
        for spot in self.parking_spots:
            if spot.is_available():
                spot.park_vehicle(vehicle)
                return True
        return False

    def un_park(self, vehicle: Vehicle) -> bool:
        for spot in self.parking_spots:
            if not spot.is_available() and spot.get_parked_vehicle() == vehicle:
                spot.unpark_vehicle()
                return True
        return False

    def display_available(self):
        print(f"Level {self.floor} status with : ")
        for spot in self.parking_spots:
            print(f"Spot {spot.get_spot_number()}: {'Available' if spot.is_available() else 'Occupied'}")

File: solutions/java/src/parkinglot/Level.java
package parkinglot;

import parkinglot.VehicleType.Vehicle;

import java.util.ArrayList;
import java.util.List;

public class Level {
    private final int floor;
    private final List<ParkingSpot> parkingSpot;

    public Level(int floor, int numberSpots) {
        this.floor = floor;
        this.parkingSpot = new ArrayList<>(numberSpots);
        for(int i = 0; i < numberSpots; i++) {
            parkingSpot.add(new ParkingSpot(i));
        }
    }

    public synchronized boolean parkVehicle(Vehicle vehicle) {
        for(ParkingSpot spot : parkingSpot) {
            if(spot.isAvailable()){
                spot.parkVehicle(vehicle);
                return true;
            }
        }
        return false;
    }

    public synchronized boolean unparkVehicle(Vehicle vehicle) {
        for (ParkingSpot spot : parkingSpot) {
            if (!spot.isAvailable() && spot.getParkedVehicle().equals(vehicle)) {
                spot.unparkVehicle();
                return true;
            }
        }
        return false;
    }

    public void displayAvailability(){
        System.out.println("Level "+ floor + " Availability");
        for(ParkingSpot spot : parkingSpot) {
            System.out.println("Spot " + spot.getSpotNumber() + ": " + (spot.isAvailable() ? "Available" : "Occupied"));
        }
    }

}
